File: dashboard/lib/operationalScope.ts
/**
 * Server-authoritative operational scope (from login /me).
 * Fallback: parse scopeJson client-side when operationalScope is absent (legacy sessions).
 */
import type { User } from '../types';
import { parseScopeDocument, type ScopeDocument } from './ownershipContext';

export interface OperationalScope {
  clusters: string[];
  namespaces: string[];
  teams: string[];
  environments: string[];
  business_services?: string[];
  crown_jewels?: string[];
  regulatory_domains?: string[];
  restricted: boolean;
}

export function emptyOperationalScope(): OperationalScope {
  return { clusters: [], namespaces: [], teams: [], environments: [], restricted: false };
}

/** Build scope from API field or scopeJson fallback. */
export function resolveOperationalScope(user: User | null | undefined): OperationalScope {
  if (!user) return emptyOperationalScope();
  const fromApi = (user as User & { operationalScope?: OperationalScope }).operationalScope;
  if (fromApi && typeof fromApi === 'object') {
    return {
      clusters: fromApi.clusters ?? [],
      namespaces: fromApi.namespaces ?? [],
      teams: fromApi.teams ?? [],
      environments: fromApi.environments ?? [],
      business_services: fromApi.business_services ?? [],
      crown_jewels: fromApi.crown_jewels ?? [],
      regulatory_domains: fromApi.regulatory_domains ?? [],
      restricted: Boolean(fromApi.restricted),
    };
  }
  return operationalScopeFromDocument(parseScopeDocument(user.scopeJson));
}

export function operationalScopeFromDocument(doc: ScopeDocument): OperationalScope {
  const clusters = [...(doc.clusters ?? []), ...(doc.cluster_ids ?? [])].filter(
    (id, i, arr) => id && arr.indexOf(id) === i,
  );
  const namespaces = doc.namespaces ?? [];
  const teams = doc.teams ?? doc.tenants ?? [];
  const environments = doc.environments ?? [];
  const business_services = doc.business_services ?? [];
  const crown_jewels = doc.crown_jewels ?? [];
  const regulatory_domains = doc.regulatory_domains ?? [];
  const restricted =
    clusters.length > 0 ||
    namespaces.length > 0 ||
    teams.length > 0 ||
    environments.length > 0 ||
    business_services.length > 0 ||
    crown_jewels.length > 0;
  return {
    clusters,
    namespaces,
    teams,
    environments,
    business_services,
    crown_jewels,
    regulatory_domains,
    restricted,
  };
}

export function scopeSummaryLabel(scope: OperationalScope, clusterName?: string): string {
  if (clusterName?.trim()) return clusterName.trim();
  if (scope.clusters.length === 1) return `cluster ${scope.clusters[0]}`;
  if (scope.clusters.length > 1) return `${scope.clusters.length} assigned clusters`;
  if (scope.restricted) return 'your assigned scope';
  return 'all monitored clusters';
}

File: core/pkg/authorization/operational_scope.go
package authorization

// OperationalScope is the server-normalized operational domain for UX (not security authority).
// Populated on login and GET /auth/me so clients do not infer scope from heuristics alone.
type OperationalScope struct {
	Clusters           []string `json:"clusters"`
	Namespaces         []string `json:"namespaces"`
	Teams              []string `json:"teams"`
	Environments       []string `json:"environments"`
	BusinessServices   []string `json:"business_services"`
	CrownJewels        []string `json:"crown_jewels"`
	RegulatoryDomains  []string `json:"regulatory_domains"`
	// Restricted is true when any explicit allow-list dimension is non-empty.
	Restricted bool `json:"restricted"`
}

// OperationalScopeFromDocument builds the API-facing scope from stored scope JSON.
func OperationalScopeFromDocument(scopeJSON string) OperationalScope {
	doc := ParseScopeDocument(scopeJSON)
	clusters := doc.clusterIDs()
	out := OperationalScope{
		Clusters:          append([]string(nil), clusters...),
		Namespaces:        append([]string(nil), doc.Namespaces...),
		Teams:             append([]string(nil), doc.Tenants...),
		Environments:      append([]string(nil), doc.Environments...),
		BusinessServices:  append([]string(nil), doc.BusinessServices...),
		CrownJewels:       append([]string(nil), doc.CrownJewels...),
		RegulatoryDomains: append([]string(nil), doc.RegulatoryDomains...),
	}
	if len(out.Clusters) > 0 || len(out.Namespaces) > 0 || len(out.Teams) > 0 || len(out.Environments) > 0 ||
		len(out.BusinessServices) > 0 || len(out.CrownJewels) > 0 || len(out.RegulatoryDomains) > 0 {
		out.Restricted = true
	}
	if len(out.Clusters) == 1 && out.Clusters[0] == "__invalid_scope__" {
		out.Clusters = nil
		out.Restricted = true
	}
	return out
}
